Appendix: Add Default Sort, Note Edit and Sort Status Enhancement into the DG (#215)
AY2324S1-CS2103T-W17-2/tp: docs/DeveloperGuide.md

* **Increased Efficiency**: This enhancement allows users to manage their
  applications more efficiently
  and save time in the event of an unintentional undos.

### Default Sort Order

**Current Implementation**: In the current version of our application, when a
user sorts by a particular field, there will be no way for the user to revert 
to the initial sort order. The only way to revert to the default sort order is 
by restarting the app.

**Planned Enhancement**: We plan to introduce a way for users to revert the list
to its default sort order. This enhancement may be particularly useful for users
who which to see the list sorted in the order they added the intern application
entries in.

**Proposed Changes**:

> * **Add new ID field into InternApplication class**: The idea behind this change 

Reviewer: ```suggestion
* **Add new `Id` field into `InternApplication` class**: The idea behind this change 
```

Reviewer: Perhaps we could standardize the notation for classes to use backticks?

Reviewer: Marked this in a few other places as well!

  is to mimic the identifier key that is present in most SQL databases. This way,
  we will be able to seamlessly sort the InternApplications according to their ID
  which is the order in which they are added.

> * **New ID Class**: The `Id` class will encapsulate the identifier key integer for

Reviewer: ```suggestion
* **New `Id` Class**: The `Id` class will encapsulate the identifier key integer for
```

  the corresponding `InternApplication`.
  The class will also have a static counter variable to count the total number of
  intern applications that is added while the app is running. This helps to ensure
  that none of the `InternApplication` objects has duplicate key values.
  `Id#generateUseableId()` is a factory method responsible for generating an ID key
  that is not used by any intern applications yet.
  It is also worth noting that the Id value will not be passed into the 
  serialised JSON format as we wish for the keys to be dynamically assigned at every 

>   run of the application. This way, it is unlikely for integer overflow in the Id 

Reviewer: ```suggestion
  run of the application. This way, it is unlikely for integer overflow in the `Id`
```

>   value.

Reviewer: Wow! That's amazing @tyouwei 👍

* **SortCommand**: The `SortCommand` class will be changed to behave differently 
  whenever the user enters `sort default` into the CommandBox. The `SortCommandParser` 
  will see that none of the prefixes are present in the input. After which, it will 
  check the preamble. If the preamble contains the string "default", the SortCommand 
  will execute the sort logic to sort by the `Id` class.

**Expected Benefits**:

> * **Enhanced Quality of Life**: Users do not have to restart the application just to see

Reviewer: ```suggestion
* **Enhanced User Experience**: Users do not have to restart the application just to see
```
Perhaps this would be a better way to phrase this?

  accordance to edit a particular note. The class will require 3 parameters
  in the constructor: The index of the application, the index of the note, 
  and the String input note that the user wants to store.

* **Additional function in InternApplication Class**: An additional function
  is required to represent the behavior that we want in note editing. Namely,
  `InternApplication#editNote()` will take in 2 parameters: The index of the
  note, and the string of input note to be stored. The function makes a copy
  of its immutable list of notes into a mutable one and swaps the `Note` in the 
  specified index input with a newly constructed `Note` from the input string.
  The function then returns a new `InternApplication` with the updated notes.

* **NoteCommandParser Changes**: The current version of our app throws an 
  exception when both `PREFIX_NOTE_DELETE` and `PREFIX_NOTE_INSERT` is detected.
  The proposed change to make way for note editing is to remove the thrown 
  exception line in place of a `NoteEditCommand` object.

**Expected Benefits**:

* **Faster note management for Users**: Editing notes will be reduced from a two

>   step process (Note deleting and then Note editing) into a one step process 

Reviewer: ```suggestion
  step process (Note deleting and then Note adding) into a one step process 
```
Do you perhaps mean Note deleting and then Note (re)adding?

Author: Oh yes, typo error

  (Note editing)

### Sort Status Enhancement

**Current Implementation**: The current implementation of sorting by status
results in a behavior where the statuses are sorted in lexicographical order.

**Planned Enhancement**: We plan to change the way Status is sorted to make
the sort logic more meaningful. (e.g. Rejected and Accepted internships are
lower than all the other statuses because they are "completed" applications
with no further follow-ups)

**Proposed Changes**:

* **Status compareTo Method**: Rather than comparing the 2 strings
  lexicographically, we compare the enum representation of the strings from both
  objects. 

* **Status StatusEnum nested class**: We propose to rearrange the enums such that
  the more important ones are above, and the less important ones are below.
  The order of importance is as follows: 

>   Offered > Assessment > Interview > Pending > Accepted > Rejected 

Reviewer: Hmmm, just a small nit, but I would argue that the order of `Status` should be kind of chronological, in a way that it follows the "stages" of an internship application process. So, it would look something like this instead:
```suggestion
  Pending > Assessment > Interview > Offered  > Accepted > Rejected 
```

**Expected Benefits**:

> * **More meaningful sort**: When Status is sorted, it will now show the more urgent

Reviewer: ```suggestion
* **More meaningful sort**: When `Status` is sorted, it will now show the more urgent
```
